pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/EquivalentStagesFinder.java

```diff
@@ -0,0 +1,339 @@
+        }
+        MailboxReceiveNode that = (MailboxReceiveNode) node2;
+        MailboxSendNode node1Sender = node1.getSender();
+        String nullSenderMessage = "This method should only be called at planning time, when the sender for a receiver "
+            + "node shall be not null.";
+        Preconditions.checkNotNull(node1Sender, nullSenderMessage);
+        MailboxSendNode node2Sender = that.getSender();
+        Preconditions.checkNotNull(node2Sender, nullSenderMessage);
+
+        // Remember that receive nodes do not have inputs. Their senders are a different attribute.
+        if (!areEquivalent(node1Sender, node2Sender)) {
+          return false;
+        }
+
+        //@formatter:off
+        return baseNode(node1, node2)
+            // Commented out fields are used in equals() method of MailboxReceiveNode but not needed for equivalence.
+            // sender stage id will be different for sure, but we want (and already did) to compare sender equivalence
+            // instead
+//          && node1.getSenderStageId() == that.getSenderStageId()
+
+            // TODO: Keys should probably be removed from the equivalence check, but would require to verify both
+            //  keys are present in the data schema. We are not doing that for now.
+            && Objects.equals(node1.getKeys(), that.getKeys())
+            // Distribution type is not needed for equivalence. We deal with difference distribution types in the
+            // spooling logic.
+//          && node1.getDistributionType() == that.getDistributionType()
+            // TODO: Sort, sort on sender and collations can probably be removed from the equivalence check, but would
+            //  require some extra checks or transformation on the spooling logic. We are not doing that for now.
+            && node1.isSort() == that.isSort()
+            && node1.isSortedOnSender() == that.isSortedOnSender()
+            && Objects.equals(node1.getCollations(), that.getCollations())
+            && node1.getExchangeType() == that.getExchangeType();
+        //@formatter:on
+      }
+
+      @Override
+      public Boolean visitFilter(FilterNode node1, PlanNode node2) {
+        if (!(node2 instanceof FilterNode)) {
+          return false;
+        }
+        FilterNode that = (FilterNode) node2;
+        return baseNode(node1, node2) && Objects.equals(node1.getCondition(), that.getCondition());
+      }
+
+      @Override
+      public Boolean visitJoin(JoinNode node1, PlanNode node2) {
+        if (!(node2 instanceof JoinNode)) {
+          return false;
+        }
+        JoinNode that = (JoinNode) node2;
+        //@formatter:off
+        return baseNode(node1, node2) && Objects.equals(node1.getJoinType(), that.getJoinType())
+            && Objects.equals(node1.getLeftKeys(), that.getLeftKeys())
+            && Objects.equals(node1.getRightKeys(), that.getRightKeys())
+            && Objects.equals(node1.getNonEquiConditions(), that.getNonEquiConditions());
```

Comment:
The `JoinStrategy` doesn't need to be compared?

Reply:
Good catch! Yes, the should, but that code was written before joinStrategy was added.